### DISCUSSION TITLE
🖼️ イメージデータの適切な置き場所について調査

### DISCUSSION THREAD

**Discussion by kagomen** [lgtm-factory/lgtm-factory#125] [Scrap]
publicとassets（または任意のディレクトリ名）の違いについて調査

**Comment by kagomen** [comment]
【別の話】【メモ】コントリビューターがLGTMデータに画像を使用する場合のディレクトリも作成した方が良さそう

**Reply by kazzyfrog** [reply]
以下のような構成でもいいかもですね！

- _dataディレクトリ：ルートに配置する、lgtm画像に関するディレクトリ
    - _data/lgtm-contents ディレクトリ: 実際にダイナミックインポートする画像のデータファイル
    - _data/asstes ディレクトリ：上記でLGTMデータに使用する画像の配置
    - README：design-guide
    - （テストファイル）：_data/lgtm-contents ディレクトリ内のファイルが、問題なく記述されているかをチェック

**Reply by kagomen** [reply]
@kazzyfrog 

公式docs一切参照せずAIに聞きながらなのでだいぶ怪しいんですが、一応個人的な結論です -> https://github.com/lgtm-factory/lgtm-factory/discussions/125#discussioncomment-10448226

LGTM Factoryがどれくらいの人に使われるかわからないので、なんとも言えないんですが、例えば1000個の画像データがpublicディレクトリ以外のディレクトリに配置されたら、ビルド時の負担が結構すごそう……？と思いました

詳しくないので、ご意見いただきたいです🙇

---

たしかにlgtm-dataより、lgtm-contentsの方が適切かも！

**Reply by kazzyfrog** [reply]
了解です！

そしたら、publicディレクトリに画像ファイル類をまとめてしまってもいいかも？ですね！

**Reply by kagomen** [reply]
了解ですー！！👍

**Reply by kagomen** [reply]
> そしたら、publicディレクトリに画像ファイル類をまとめてしまってもいいかも？ですね！

あ、これって、lgtmデータに使われる画像データも、背景画像やロゴ画像もpublicディレクトリに配置する、ということで合ってますか？

**Reply by kazzyfrog** [reply]
あってます！
自分的には、
publicディレクトリ内で、サブディレクトリで分けるような感じが、明確かな？と思った感じです！
https://nextjs.org/docs/app/building-your-application/optimizing/static-assets

とはいえ、ルートのディレクトリ自体で、publicと、asstesに分けても、全然良いかと思います！

**Comment by kagomen** [comment]
Next.jsでの使用素材の配置場所として、`public` ディレクトリと `assets` ディレクトリ（または類似の名前のディレクトリ）を使用する場合、それぞれに特徴があります。以下にその違いを説明します：

1. `public` ディレクトリ

- 特徴:
  - Next.jsによって自動的に認識され、ルートURL（`/`）から直接アクセス可能
  - ビルド時に最適化されない
  - 常に同じファイル名で参照される

- 使用例:
  - ファビコン
  - robots.txt
  - サイトマップ
  - 静的な画像ファイル（変更されないもの）

- 参照方法:
  ```jsx
  <img src="/image.png" alt="My Image" />
  ```

2. `assets` ディレクトリ（または類似の名前）

- 特徴:
  - プロジェクト構造をより柔軟に管理できる
  - ビルド時に最適化される可能性がある（設定による）
  - インポートして使用するため、ファイル名の変更やパスの移動に強い

- 使用例:
  - コンポーネント固有の画像
  - SVGファイル（インラインで使用する場合）
  - スタイルシート
  - その他のアセットファイル

- 参照方法:
  ```jsx
  import myImage from '../assets/images/myImage.png';
  
  <img src={myImage} alt="My Image" />
  ```

主な違い:

1. アクセス方法:
   - `public`: URLから直接アクセス可能
   - `assets`: インポートして使用

2. 最適化:
   - `public`: 最適化されない
   - `assets`: ビルドプロセスで最適化される可能性がある

3. 柔軟性:
   - `public`: ファイル名の変更に弱い
   - `assets`: インポートパスを更新すれば柔軟に対応可能

4. 用途:
   - `public`: 静的ファイル、メタファイルに適している
   - `assets`: コンポーネントや機能に関連するファイルに適している

選択の基準:
- 静的でグローバルに使用するファイル → `public`
- コンポーネントや機能に紐づくファイル、最適化が必要なファイル → `assets`

両方のアプローチを状況に応じて使い分けることで、効率的なプロジェクト構造を維持できます。

**Reply by kagomen** [reply]
サイトの背景画像とロゴのイメージデータについて、それぞれの適切な配置場所を検討してみましょう。

1. サイトの背景画像:

一般的には `assets` ディレクトリ（または類似の名前のディレクトリ）が適しています。

理由:
- 背景画像はサイトのデザインの一部であり、頻繁に変更される可能性があります。
- CSSでスタイリングする際に、インポートして使用するのが便利です。
- ビルド時の最適化プロセスの恩恵を受けられます（画像の圧縮など）。

例:
```jsx
import bgImage from '../assets/images/background.jpg';

const styles = {
  background: `url(${bgImage}) no-repeat center center fixed`,
  backgroundSize: 'cover',
};

function Layout({ children }) {
  return <div style={styles}>{children}</div>;
}
```

2. ロゴのイメージデータ:

ロゴについては、状況によって `public` ディレクトリと `assets` ディレクトリのどちらも適している場合があります。

a) `public` ディレクトリが適している場合:
- ロゴが固定で、長期間変更されない場合
- SEOの観点から、ロゴのURLを固定したい場合
- 外部から直接アクセスする必要がある場合（OGイメージとしても使用する場合など）

例:
```jsx
<img src="/logo.png" alt="Company Logo" />
```

b) `assets` ディレクトリが適している場合:
- ロゴが頻繁に更新される可能性がある場合
- 異なるバージョンのロゴ（ダークモード用など）を管理する場合
- コンポーネントと密接に関連している場合

例:
```jsx
import logo from '../assets/images/logo.png';

function Header() {
  return <img src={logo} alt="Company Logo" />;
}
```

結論:

1. サイトの背景画像: `assets` ディレクトリを推奨します。

2. ロゴのイメージデータ: 
   - 固定で変更頻度が低い場合は `public` ディレクトリ
   - 変更頻度が高い、または複数バージョンを管理する場合は `assets` ディレクトリ

ただし、プロジェクトの要件や開発チームの方針によって最適な選択は変わる可能性があります。一貫性のある方針を決めて、チーム内で統一することが重要です。

**Reply by kagomen** [reply]
ご質問ありがとうございます。背景画像やロゴ画像が滅多に変更されない場合、確かに配置の選択肢が変わってきます。このような状況では、`public` ディレクトリを使用することも十分に適切な選択肢となります。

背景画像を `public` ディレクトリに配置することのメリット：

1. 直接URLでアクセス可能：CDNやキャッシュの設定が容易になります。

2. シンプルな参照：コード内で直接パスを指定できるため、シンプルです。

3. ビルド時間の短縮：`assets` ディレクトリ内のファイルはビルド時に処理されますが、`public` 内のファイルはそのままコピーされるため、ビルドが高速化される可能性があります。

4. バージョン管理のしやすさ：変更頻度が低いファイルを分離して管理できます。

使用例：
```css
body {
  background-image: url('/images/background.jpg');
}
```

```jsx
<img src="/images/logo.png" alt="Company Logo" />
```

ただし、以下の点に注意が必要です：

1. 画像の最適化：`public` ディレクトリ内のファイルは自動的に最適化されないため、必要に応じて手動で最適化する必要があります。

2. ファイル名の一貫性：`public` ディレクトリ内のファイル名を変更すると、参照しているすべての箇所を更新する必要があります。

結論：
背景画像とロゴ画像が滅多に変更されない場合、`public` ディレクトリに配置することも十分に適切な選択肢です。特に、これらの画像が静的で、サイト全体で一貫して使用される場合はなおさらです。

ただし、プロジェクトの規模や要件、チームの方針によっては、依然として `assets` ディレクトリを使用する利点もあります（例：ビルドプロセスでの自動最適化、モジュール化されたインポートなど）。

最終的には、プロジェクトの具体的な要件とチームの方針に基づいて判断することが重要です。一貫性のあるアプローチを選択し、チーム全体でそれを維持することが、長期的なメンテナンス性の向上につながります。

**Reply by kagomen** [reply]
素晴らしい質問です。「最適化」と「キャッシュ」について詳しく説明させていただきます。

1. 「最適化」について：

画像の最適化とは、主に以下のような処理を指します：

a) 圧縮：
   - ファイルサイズを削減しつつ、視覚的な品質をできるだけ維持します。
   - JPEGやPNGの圧縮レベルの調整、不要なメタデータの削除などが含まれます。

b) リサイズ：
   - 必要以上に大きな画像を適切なサイズに縮小します。

c) フォーマット変換：
   - より効率的な形式（例：WebPやAVIF）への変換を行います。

d) レスポンシブ画像：
   - 異なる画面サイズやデバイスに対応するための複数バージョンの生成。

e) 遅延読み込み：
   - 画像の読み込みを必要なタイミングまで遅らせる設定。

Next.jsのビルドプロセスでは、`next/image` コンポーネントを使用すると、これらの最適化の多くが自動的に行われます。一方、`public` ディレクトリ内の画像は、このような自動最適化プロセスを通過しません。

2. キャッシュと画像の更新について：

`public` ディレクトリ内の画像ファイルを変更した場合：

a) ブラウザキャッシュ：
   - ユーザーのブラウザにキャッシュされている場合、変更が即時に反映されない可能性があります。

b) CDNキャッシュ：
   - CDNを使用している場合、CDN側でもキャッシュが効いている可能性があります。

c) サーバーキャッシュ：
   - サーバー側でもキャッシュが設定されている場合があります。

これらのキャッシュ問題に対処するための方法：

1. ファイル名の変更：
   新しいバージョンの画像には異なるファイル名を使用します。
   例：`logo.png` → `logo-v2.png`

2. クエリパラメータの追加：
   ファイル名は変更せず、URLにバージョンやタイムスタンプを追加します。
   例：`/images/logo.png?v=2` または `/images/logo.png?t=1234567890`

3. キャッシュヘッダーの設定：
   サーバー側でキャッシュの有効期限や条件を適切に設定します。

4. キャッシュバスティング：
   ビルドプロセスで自動的にファイル名やパスにハッシュを追加する技術。

5. 手動でのキャッシュクリア：
   CDNやサーバーのキャッシュを手動でクリアします。

Next.jsでは、デフォルトで一部のキャッシュ対策が実装されていますが、`public` ディレクトリの静的アセットに関しては、開発者が適切な戦略を選択する必要があります。

結論として、`public` ディレクトリの画像を更新する際は、キャッシュの影響を考慮し、必要に応じて上記のような対策を講じることが重要です。一方で、`next/image` コンポーネントを使用すると、多くの最適化とキャッシュ管理が自動的に行われるため、管理が容易になる利点があります。

**Reply by kagomen** [reply]
背景画像を `assets` フォルダに配置し、Next.jsの最適化を受けるには、いくつかの方法があります。以下に主な方法を説明します：

1. CSS-in-JSを使用する方法

Next.jsの `next/image` コンポーネントを直接背景画像として使用することはできませんが、CSS-in-JSライブラリ（例：styled-components, Emotion）を使用して最適化された画像URLを背景に設定することができます。

```jsx
import Image from 'next/image';
import styled from 'styled-components';
import backgroundImage from '../assets/images/background.jpg';

const StyledBackground = styled.div`
  background-image: url(${props => props.src});
  background-size: cover;
  background-position: center;
  width: 100%;
  height: 100vh;
`;

function BackgroundComponent() {
  return (
    <>
      <Image
        src={backgroundImage}
        layout="fill"
        objectFit="cover"
        alt="background"
        quality={100}
      />
      <StyledBackground src={backgroundImage.src} />
    </>
  );
}
```

この方法では、`next/image` コンポーネントが画像を最適化し、その最適化されたURLを背景画像として使用します。

2. getStaticProps を使用する方法

`getStaticProps` を使用して、ビルド時に画像のパスを取得し、最適化された画像URLを生成することができます。

```jsx
import { getStaticProps } from 'next';
import Image from 'next/image';

export default function Home({ optimizedImageUrl }) {
  return (
    <div style={{
      backgroundImage: `url(${optimizedImageUrl})`,
      backgroundSize: 'cover',
      backgroundPosition: 'center',
      width: '100%',
      height: '100vh'
    }}>
      <h1>Welcome to my site</h1>
    </div>
  );
}

export async function getStaticProps() {
  const optimizedImageUrl = require(`../assets/images/background.jpg?url`);
  return {
    props: {
      optimizedImageUrl
    }
  };
}
```

この方法では、ビルド時に画像が最適化され、その最適化されたURLがコンポーネントに渡されます。

3. next.config.js でカスタム設定を行う方法

`next.config.js` ファイルで `webpack` の設定をカスタマイズし、画像の最適化を行うことができます。

```javascript
const path = require('path');

module.exports = {
  webpack(config) {
    config.module.rules.push({
      test: /\.(png|jpg|gif|svg)$/,
      use: [
        {
          loader: 'url-loader',
          options: {
            limit: 8192,
            publicPath: '/_next/static/images/',
            outputPath: 'static/images/',
            name: '[name]-[hash].[ext]',
          },
        },
      ],
    });
    return config;
  },
};
```

この設定により、画像ファイルが最適化され、適切なパスに出力されます。

使用例：

```jsx
import backgroundImage from '../assets/images/background.jpg';

function BackgroundComponent() {
  return (
    <div style={{
      backgroundImage: `url(${backgroundImage})`,
      backgroundSize: 'cover',
      backgroundPosition: 'center',
      width: '100%',
      height: '100vh'
    }}>
      <h1>Welcome to my site</h1>
    </div>
  );
}
```

これらの方法を使用することで、`assets` フォルダに配置した背景画像に対してもNext.jsの最適化機能を活用することができます。プロジェクトの要件や構造に応じて、最適な方法を選択してください。

**Comment by kagomen** [comment]
まとめ：

画像データは**可能であれば**assetsディレクトリに置く方が、よい
- assetsに置くと、ハッシュが付与されるため、キャッシュ更新の複雑さを避けられる
- assetsに置くと画像最適化の恩恵を受けられる可能性がある

ただし、assetsに置くと、ビルド時に毎回画像データが読み込まれる
ので、サーバーに負担がかかり、ビルドの時間が遅くなる（＆、無料枠を食いつぶしたりする？）
一方、画像データをpublicディレクトリに置いておくと、静的ファイルとして扱われるので、ビルド時にスルーされ、これらの問題は解消される

結論：

サイトに使う画像データは、（数も少ないし）assetsディレクトリに置くのが開発時に使いやすくて良さそう
コントリビューターがLGTMデータに使う画像データは、publicディレクトリに置くのが良さそう

**Comment by kagomen** [comment]
## public×css
4.4MB
public, max-age=0

## public×next/image
102kB
public, max-age=0, must-revalidate

## assets×next/image
102kB
public, max-age=315360000, immutable

**Reply by kagomen** [reply]
`placeholder="blur"`は、import したファイルだと`blurDataUrl`を指定しなくても使える

https://nextjs.org/docs/app/api-reference/components/image#placeholder

**Reply by kagomen** [reply]
1. 前者 (`public, max-age=0, must-revalidate`)
   - ページをリロードすると、ブラウザは必ずサーバーに新しいバージョンがあるかどうかを確認します。
   - 技術的には「キャッシュが即座に期限切れになる」という状態です。
   - しかし、サーバーが「304 Not Modified」を返せば、キャッシュされたバージョンを使用できます。
   - つまり、毎回サーバーとの通信は発生しますが、必ずしも新しいコンテンツをダウンロードするわけではありません。

2. 後者 (`public, max-age=315360000, immutable`)
   - ページをリロードしても、指定された期間（この場合約10年間）はブラウザがキャッシュを使い続けます。
   - サーバーへの確認すら行われません。
   - ブラウザのキャッシュをクリアしない限り、同じバージョンが使用され続けます。

重要な点：
- 前者はコンテンツが頻繁に更新される可能性がある場合に使用します。
- 後者は、コンテンツが絶対に変更されない（例：特定バージョンのJSライブラリ）場合に使用します。

したがって、「リロードしたらキャッシュが切れる」というよりも、前者は「リロードするたびにサーバーに確認する」、後者は「リロードしてもキャッシュを信頼し続ける」と表現するのがより正確です。

**Reply by kagomen** [reply]
Next.jsの場合、特にVercelなどのサービスを使用してデプロイする場合、後者のキャッシュ設定（`public, max-age=315360000, immutable`）は以下のような挙動をします：

1. ビルド時の動作:
   - 新しいビルドが実行されると、静的アセット（JS、CSS、画像ファイルなど）には通常、新しいファイル名やパスが割り当てられます。
   - これは通常、ファイル名にハッシュを含めることで実現されます（例：`main-f7e9b.js`）。

2. デプロイ後の挙動:
   - 新しいビルドがデプロイされると、これらの新しい名前のファイルが新しいURLで提供されます。
   - 古いファイルは異なるURLにあるため、新しいデプロイメントの影響を受けません。

3. ブラウザの挙動:
   - ユーザーが更新されたサイトにアクセスすると、新しいHTMLが読み込まれ、そこには新しいアセットファイルへのリンクが含まれています。
   - ブラウザはこれらの新しいURLを初めて見るので、新しいファイルをダウンロードします。
   - 古いファイルは異なるURLにキャッシュされているため、影響を受けません。

4. キャッシュの更新:
   - 実質的に、各ビルドで新しいアセットが生成されるため、長期キャッシュ（`immutable`）を使用しても問題ありません。
   - 古いキャッシュは、新しいHTMLが参照しなくなるため、自然に「廃れて」いきます。

5. CDNの挙動:
   - CDNも同様に、新しいファイルを新しいURLで提供し始めます。
   - 古いファイルは異なるURLにあるため、CDN上でも問題なくキャッシュされ続けます。

したがって、GitHubにプッシュしてビルドが走った場合：

- 新しいアセットファイルが生成され、デプロイされます。
- 古いファイルのキャッシュは技術的には「更新されない」ですが、新しいHTMLが新しいファイルを参照するため、実質的には新しいコンテンツが提供されます。
- ユーザーは自動的に新しいバージョンのサイトを見ることになります。

この方法により、長期キャッシュの利点（高速な読み込み、サーバー負荷の軽減）を維持しながら、コンテンツの更新を確実に行うことができます。
